### PR TITLE
Handle nil cask urls caused by unsupported macOS version

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

The goal here is to handle the case where a cask might have a nil url stanza because that cask is not available on the current version of macOS or the given architecture. This just moves those checks from the end of the `Cask::Installer#fetch` method to the beginning so that we don't try and download casks that are missing urls. A similar change was also applied to the `brew fetch` command.

This will now provide a helpful error message like so:
```
Error: This software does not run on macOS versions older than Big Sur.
```

Beyond that it no longer tries to run the url stanza with a nil value when loading casks from the API.

### CODE REVIEW

File: Library/Homebrew/cask/installer.rb [hunk 1]
```diff
@@ -315,10 +309,13 @@ def missing_cask_and_formula_dependencies
       end
     end
 
-    def cask_and_formula_dependencies
+    # TODO: move dependencies to a separate class,
+    #       dependencies should also apply for `brew cask stage`,
+    #       override dependencies with `--force` or perhaps `--force-deps`
```
Comment (apainintheneck) [inline Library/Homebrew/cask/installer.rb:314]: Note to self: this comment should either get update or removed. `brew cask stage` is not longer a valid Homebrew command.

File: Library/Homebrew/cask/cask_loader.rb [hunk 2]
```diff
@@ -277,7 +277,7 @@ def load(config:)
             sha256 json_cask[:sha256]
           end
 
-          url json_cask[:url], **json_cask.fetch(:url_specs, {})
+          url json_cask[:url], **json_cask.fetch(:url_specs, {}) if json_cask[:url].present?
```
Comment (apainintheneck) [inline Library/Homebrew/cask/cask_loader.rb:280]: Before:
```
/u/l/Homebrew (master|✔) [1]$ brew info gpgfrontend
Error: no implicit conversion of nil into Hash
/usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:280:in `block in load'
/usr/local/Homebrew/Library/Homebrew/cask/cask.rb:102:in `instance_eval'
/usr/local/Homebrew/Library/Homebrew/cask/cask.rb:102:in `refresh'
/usr/local/Homebrew/Library/Homebrew/cask/cask.rb:95:in `config='
/usr/local/Homebrew/Library/Homebrew/cask/cask.rb:74:in `initialize'
...
/usr/local/Homebrew/Library/Homebrew/cmd/info.rb:112:in `info'
/usr/local/Homebrew/Library/Homebrew/brew.rb:94:in `<main>'
```

After:
```
/u/l/Homebrew (handle-nil-urls-in-cask-installer|✔) $ brew info gpgfrontend
==> gpgfrontend: 2.1.1
https://gpgfrontend.bktus.com/
Not installed
From: https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/g/gpgfrontend.rb
==> Name
GpgFrontend
==> Description
OpenPGP/GnuPG crypto, sign and key management tool
==> Artifacts
GpgFrontend.app (App)
==> Analytics
install: 32 (30 days), 108 (90 days), 210 (365 days)
```

File: Library/Homebrew/cask/download.rb [hunk 3]
```diff
@@ -24,6 +24,8 @@ def initialize(cask, quarantine: nil)
 
     sig { override.returns(T.nilable(::URL)) }
     def url
+      return if cask.url.nil?
```
Comment (apainintheneck) [inline Library/Homebrew/cask/download.rb:27]: Small thing that will give us better error messages in the future if we try downloading casks without urls.

File: Library/Homebrew/cask/installer.rb [hunk 4]
```diff
@@ -64,12 +64,12 @@ def fetch(quiet: nil, timeout: nil)
       odebug "Cask::Installer#fetch"
 
       load_cask_from_source_api! if @cask.loaded_from_api? && @cask.caskfile_only?
-
       verify_has_sha if require_sha? && !force?
+      check_requirements
```
Comment (apainintheneck) [inline Library/Homebrew/cask/installer.rb:68]: Before:
```
/u/l/Homebrew (master|✔) $ brew install gpgfrontend
Error: undefined method `specs' for nil:NilClass
/usr/local/Homebrew/Library/Homebrew/cask/download.rb:27:in `url'
/usr/local/Homebrew/Library/Homebrew/cask/download.rb:104:in `determine_url'
/usr/local/Homebrew/Library/Homebrew/downloadable.rb:126:in `determine_url_mirrors'
/usr/local/Homebrew/Library/Homebrew/downloadable.rb:74:in `downloader'
/usr/local/Homebrew/Library/Homebrew/downloadable.rb:87:in `fetch'
/usr/local/Homebrew/Library/Homebrew/cask/download.rb:51:in `fetch'
/usr/local/Homebrew/Library/Homebrew/cask/installer.rb:167:in `download'
/usr/local/Homebrew/Library/Homebrew/cask/installer.rb:70:in `fetch'
/usr/local/Homebrew/Library/Homebrew/cask/installer.rb:99:in `install'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:244:in `block in install'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:233:in `each'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:233:in `install'
/usr/local/Homebrew/Library/Homebrew/brew.rb:94:in `<main>'
```
(obviously without the Sorbet lines in the stack trace)

After:
```
/u/l/Homebrew (handle-nil-urls-in-cask-installer|✔) [1]$ brew install gpgfrontend
Error: This software does not run on macOS versions older than Big Sur.
```

File: Library/Homebrew/cmd/fetch.rb [hunk 5]
```diff
@@ -167,6 +167,11 @@ def self.fetch
           SimulateSystem.with os: os, arch: arch do
             cask = Cask::CaskLoader.load(ref)
 
+            if cask.depends_on.macos&.satisfied? == false
+              opoo "#{cask.token}: #{cask.depends_on.macos.message(type: :cask)}"
+              next
+            end
```
Comment (apainintheneck) [inline Library/Homebrew/cmd/fetch.rb:173]: Command: `brew fetch gpgfrontend`

Before:
```
Error: undefined method `specs' for nil:NilClass
/usr/local/Homebrew/Library/Homebrew/cask/download.rb:27:in `url'
/usr/local/Homebrew/Library/Homebrew/cask/download.rb:104:in `determine_url'
/usr/local/Homebrew/Library/Homebrew/downloadable.rb:126:in `determine_url_mirrors'
/usr/local/Homebrew/Library/Homebrew/downloadable.rb:74:in `downloader'
/usr/local/Homebrew/Library/Homebrew/downloadable.rb:50:in `cached_download'
/usr/local/Homebrew/Library/Homebrew/cmd/fetch.rb:232:in `fetch_fetchable'
/usr/local/Homebrew/Library/Homebrew/cmd/fetch.rb:197:in `fetch_cask'
/usr/local/Homebrew/Library/Homebrew/cmd/fetch.rb:174:in `block (3 levels) in fetch'
/usr/local/Homebrew/Library/Homebrew/simulate_system.rb:29:in `with'
/usr/local/Homebrew/Library/Homebrew/cmd/fetch.rb:167:in `block (2 levels) in fetch'
/usr/local/Homebrew/Library/Homebrew/cmd/fetch.rb:164:in `each'
/usr/local/Homebrew/Library/Homebrew/cmd/fetch.rb:164:in `block in fetch'
/usr/local/Homebrew/Library/Homebrew/cmd/fetch.rb:94:in `each'
/usr/local/Homebrew/Library/Homebrew/cmd/fetch.rb:94:in `fetch'
/usr/local/Homebrew/Library/Homebrew/brew.rb:94:in `<main>'
```
(without the Sorbet stack trace lines)

After:
```
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
```

File: Library/Homebrew/cask/installer.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cask/installer.rb:314]: Remove I reckon.

Comment (MikeMcQuaid) [reply Library/Homebrew/cask/installer.rb:314]:
```suggestion
```

**Comment by EricFromCanada** [comment]
There's a similar issue in `brew audit --cask` for casks that don't have an `on_<system>` block for the current system, e.g.:
```
$ brew audit --skip-style --except=version homebrew/cask/apparency -d
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::TapLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Formula/apparency.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromTapLoader): loading homebrew/cask/apparency
==> Auditing Cask apparency on os high_sierra and arch intel
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromTapPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Casks/a/apparency.rb
==> Auditing pkg stanza: allow_untrusted
==> Auditing stanzas which require an uninstall
==> Auditing preflight and postflight stanzas
==> Auditing single uninstall_* and zap stanzas
==> Auditing required stanzas
==> undefined method `latest?' for nil:NilClass
/usr/local/Homebrew/Library/Homebrew/cask/audit.rb:270:in `audit_latest_with_livecheck'
...
```

**Comment by apainintheneck** [comment]
@EricFromCanada Good catch! I was wondering if this would affect any other part of the codebase but I couldn't find anything with a quick search. I can try and make a PR for that if you're not already working on it.

**Comment by EricFromCanada** [comment]
Go ahead, I've got other metaphorical fish to fry.

**Comment by apainintheneck** [comment]
It looks like `MacOSRequirement` and `ArchRequirement` do not respect `SimulateSystem` which means this isn't a super easy, one line fix. It also means that the changes to `brew fetch` made in this PR won't work as intended when users are simulating other os/arch combinations but I guess it never really worked as intended anyway. I'm not sure how common of a use case that was before though.

```
/u/l/Homebrew (master|✔) $ brew fetch --os=all gpgfrontend -d
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading gpgfrontend
Warning: gpgfrontend: This software does not run on macOS versions older than Big Sur.
```

**Comment by apainintheneck** [comment]
On second thought, we could probably just skip all os/arch combinations that have nil cask urls. I'm not sure if this affects anything else though.

**Comment by MikeMcQuaid** [comment]
> It looks like `MacOSRequirement` and `ArchRequirement` do not respect `SimulateSystem` which means this isn't a super easy, one line fix.

@apainintheneck could you open up a `help wanted` issue to get this fixed?

**Comment by apainintheneck** [comment]
@MikeMcQuaid I'm not sure it is needed for most use cases or if we want requirements to check the simulated system instead of the real one.

@Rylan12 @reitermarkus Do you have any thoughts on this? I seem to remember you've both worked a bit on this part of the codebase before.

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid I'm not sure it is needed for most use cases or if we want requirements to check the simulated system instead of the real one.

I feel like the simulated system should be checked, yeh.

**Pull request merged**